### ISSUE TITLE
Reproducing problem on cityscapes

### ISSUE THREAD

**Issue by FishYuLi** [junyanz/pytorch-CycleGAN-and-pix2pix#150]
Hello! I have read your paper very carefully and tried to reproduce your experiments on cityscapes dataset (photo to label). I changed the input size from 256 to 128 and use resnet_6blocks for G as stated in the appendix, and trained a model, and evaluated the performance of photo to label generator. You got (0.58, 0.22, 0.16) results for pixel acc, cls acc, and iou. But my best results are (0.51, 0.16, 0.10), which have quiet a large margin. I wondered if there are any other details that I should change. What is your configs for cityscapes? And can the size of batch_size produce an influence on the final results? Also, how did you evaluate the segmentation results of 128x128 generated images? Did you resize the original label image to 128x128 and do the evaluation? Thanks a lot. This work is amazing.

**Comment by junyanz** [comment]
We use `batchSize=1` and we got worse results with larger batchSize.  I remembered a few details (e.g.  saving the output as png images, `loadSize=143`, `fineSize=128`). We use the evaluation [code]((https://github.com/phillipi/pix2pix/tree/master/scripts/eval_cityscapes) in pix2pix. See more discussion [here(https://github.com/phillipi/pix2pix/issues/112).

**Comment by FishYuLi** [comment]
@junyanz Ok, I see. Thanks a lot! I use `batchSize=20` on 4 gpus to accelerate training. I will train the model again and try your evaluation code in pix2pix.

**Comment by FishYuLi** [comment]
Sorry it's me again. I tried your evaluation [code](https://github.com/phillipi/pix2pix/tree/master/scripts/eval_cityscapes). It looks normal when I take the fake image as input, but it's so weird that when the input is the true image, the segmentation becomes so bad. My question is,
1. what iou does this of this caffe model get on cityscapes?
2. is there any special configs that we need to change when running normal images?
![results](https://user-images.githubusercontent.com/17564896/32892903-ac99332e-cb12-11e7-9023-a3b9d51c0615.png)
The numbers are (pixel acc., cls acc., iou).

**Comment by tinghuiz** [comment]
Please see the notes from the original pix2pix torch repo (copied below): 
The pre-trained model does not work well on Cityscapes in the original resolution (1024x2048) as it was trained on 256x256 images that are resized to 1024x2048. The purpose of the resizing was to 1) keep the label maps in the original high resolution untouched and 2) avoid the need of changing the standard FCN training code for Cityscapes. To get the ground-truth numbers in the paper, you need to resize the original Cityscapes images to 256x256 before running the evaluation code.

**Comment by junyanz** [comment]
Please see this [discussion](https://github.com/phillipi/pix2pix/issues/116). Maybe @tinghuiz has more insights.

**Comment by lx7555** [comment]
非常感谢回复。
1.我目前已经将cityscapes数据集中自带的scripts下载好，我想这样就解决了request "labels"的问题。
2,我已经下载好cityscapes数据集原图(leftImg8bit)和精细标注(gtFine_trainvaltest)。请问如何安排和放置文件夹的位置，能够让评估程序读取到数据集的信息？
3.你说通过将cityscapes数据集压缩成256*256再放大成1024*2048，再将某一行代码取消注释，可以评估cityscapes数据集的原始像素图像。原文说只能评估256*256，请问你取消代码注释后。评估cityscapes数据集原始像素图像效果如何？为什么要先裁剪，后还原？？
谢谢

**Comment by lx7555** [comment]
I want to evaluate cityscapes datasets,because I try to reproduce the fcn score results from pip2pip by using https://github.com/junyanz/pytorch-CycleGAN-and-pix2pix code.
I try many ways.I can not get any results when I run the scripts. All parameters are zero.
I still want to know how to configure the cityspaces dataset
There are three folders with gtFine, originals image, and predictions.
Are the gtFine and predictions color or grayscale? And what is the size of these three types of pictures?
I use python2,caffe.I Is that possible that it was because i was using python2.7 but not python3 ?
@FishYuLi

**Comment by junyanz** [comment]
@tinghuiz

**Comment by ZhangCZhen** [comment]
Hello, I evaluated the results after using the cityscape dataset, and the values obtained are far from those mentioned in the paper. So I would like to ask, is the code given by 'evaluate.py' only used to evaluate the results of label2image? Can I evaluate image2label? After I use RGB image to generate semantic segmentation results, how do I evaluate the segmentation results? Thanks a lot!

**Comment by FishYuLi** [comment]
> Hello, I evaluated the results after using the cityscape dataset, and the values obtained are far from those mentioned in the paper. So I would like to ask, is the code given by 'evaluate.py' only used to evaluate the results of label2image? Can I evaluate image2label? After I use RGB image to generate semantic segmentation results, how do I evaluate the segmentation results? Thanks a lot!

@ZhangCZhen 
1. Yes, the given 'evaluate.py' is for label2image. It sends the generated image to a FCN model to generate a mask, and evaluates the mask with ground truth mask.
2. Of couse you can use it to evaluate  image2label. Just directly load your generated label, transfer the RGB label to $h\times w \times c$, and you can re-use the evaluation code in 'evaluate.py'.

**Comment by ZhangCZhen** [comment]
@FishYuLi    Thanks for your reply. 
1. When you use 'evaluate.py' to evaluate the label2RGB image results, did you get values similar to those mentioned in the paper? I used the RGB images in the dataset for testing, and the accuracy is still poor. 
2. You mean I use the 'evaluate.py' to evaluate image2label results, I don't need the FCN model?

**Comment by FishYuLi** [comment]
@ZhangCZhen 
1. Yes, I got almost the same FCN score.
2. Yes, FCN model is used to transfer the generated real image to label again so that we can compare it with GT masks. Please refer to the paper of Pix2pix carefully to see how exactly the evaluation metric is designed.

**Comment by ZhangCZhen** [comment]
@FishYuLi 
Thank you for your reply, you are really enthusiastic！！

**Comment by ZhangCZhen** [comment]
@FishYuLi 
Sorry it's me again. When you evaluate the results of image2label, you need to convert the semantic segmentation results into Ids. Right? How do you convert the semantically segmented RGB map to Ids map?  I mean, the pixel values of the color label image you got do not exactly match the pixel values in 'labels.py'. How do you evaluate the semantic segmentation results?  Thanks a lot. 
   I will be very grateful if you could send me your evaluation code. My email: 825762985@qq.com

**Comment by FishYuLi** [comment]
@ZhangCZhen 
In my case, I first got a standard color map for Cityscapes dataset. (I mean you should get what is the exact RGB color for a certain category in ground truth label map.) Then, for the generated label, calculate the L2 distance of each pixel with the RGB color of all categories, and asign that pixel to the category with the smallest L2 distance. Then you get the WxHxC prediction.

**Comment by ZhangCZhen** [comment]
@FishYuLi 
With your help, I have reproduced the evaluation results of label2image, and I am very grateful for this. Thank you very much! But I don't understand your above method of evaluating image2label. I have a few questions for this to ask you.
1. Did you use 'evaluate.py' when evaluating image2label?   I think you used it, did you?
2. If you used 'evaluate.py' to evaluate the generated results of image2label, is the ground truth '_gtFine_labelIds.png'?  The calculation of accuracy values during the evaluation process uses the 'trainIds' of the ground truth. (Use 'assign_trainIds' function in 'cityscape.py' to align).  But I don't know how to convert the generated color label map to trainIds value.  
3. "Of couse you can use it to evaluate image2label. Just directly load your generated label, transfer the RGB label to H x W x C, and you can re-use the evaluation code in 'evaluate.py'." --This is your reply. I don't know how to transfer the RGB label to H*W*C, and why you transfer it?
Looking forward to your reply. Thanks a lot!

**Comment by CR-Gjx** [comment]
> @junyanz Ok, I see. Thanks a lot! I use `batchSize=20` on 4 gpus to accelerate training. I will train the model again and try your evaluation code in pix2pix.

Hi, did you re-train the model and get the similar results with them in the paper? Could you provide some details (e.g parameters) for me? Now I cannot reproduce the results using default parameters. Thanks! @FishYuLi

**Comment by FishYuLi** [comment]
> > @junyanz Ok, I see. Thanks a lot! I use `batchSize=20` on 4 gpus to accelerate training. I will train the model again and try your evaluation code in pix2pix.
> 
> Hi, did you re-train the model and get the similar results with them in the paper? Could you provide some details (e.g parameters) for me? Now I cannot reproduce the results using default parameters. Thanks! @FishYuLi

Yes, I got similar results as that in the paper just with the default parameters. I think you may try to set your `batchSize=1` on single gpu. I have also tried to enlarge the batchSize but find that the model is very sensitive to batch size. Results become worse with large batch size. Also note that the input size should be 128x128 but not 256.

**Comment by CR-Gjx** [comment]
Thanks for your help, I have reproduced the results
